File: src/FScript.LanguageServer/LspServer.fs
namespace FScript.LanguageServer

open System
open System.Text.Json.Nodes

module LspServer =
    open LspModel

    let run () =
        let mutable keepRunning = true
        let mutable shutdownReceived = false

        while keepRunning do
            match LspProtocol.tryReadMessage () with
            | None -> keepRunning <- false
            | Some payload ->
                try
                    match JsonNode.Parse(payload) with
                    | null -> ()
                    | rootNode ->
                        match asObject rootNode with
                        | None -> ()
                        | Some root ->
                            let methodName = tryGetString root "method"
                            let idNode = tryGetNode root "id"
                            let paramsObj = tryGetObject root "params"

                            match methodName, idNode, paramsObj with
                            | Some "initialize", Some idNode, paramsObj ->
                                LspHandlers.handleInitialize idNode paramsObj
                            | Some "initialized", _, _ ->
                                ()
                            | Some "shutdown", Some idNode, _ ->
                                shutdownReceived <- true
                                LspProtocol.sendResponse idNode None
                            | Some "exit", _, _ ->
                                keepRunning <- false
                                if shutdownReceived then
                                    Environment.ExitCode <- 0
                                else
                                    Environment.ExitCode <- 1
                            | Some "textDocument/didOpen", _, Some paramsObj ->
                                LspHandlers.handleDidOpen paramsObj
                            | Some "textDocument/didChange", _, Some paramsObj ->
                                LspHandlers.handleDidChange paramsObj
                            | Some "textDocument/didClose", _, Some paramsObj ->
                                LspHandlers.handleDidClose paramsObj
                            | Some "textDocument/completion", Some idNode, Some paramsObj ->
                                LspHandlers.handleCompletion idNode paramsObj
                            | Some "textDocument/semanticTokens/full", Some idNode, Some paramsObj ->
                                LspHandlers.handleSemanticTokens idNode paramsObj
                            | Some "textDocument/hover", Some idNode, Some paramsObj ->
                                LspHandlers.handleHover idNode paramsObj
                            | Some "textDocument/definition", Some idNode, Some paramsObj ->
                                LspHandlers.handleDefinition idNode paramsObj
                            | Some "textDocument/typeDefinition", Some idNode, Some paramsObj ->
                                LspHandlers.handleTypeDefinition idNode paramsObj
                            | Some "textDocument/documentSymbol", Some idNode, Some paramsObj ->
                                LspHandlers.handleDocumentSymbol idNode paramsObj
                            | Some "textDocument/references", Some idNode, Some paramsObj ->
                                LspHandlers.handleReferences idNode paramsObj
                            | Some "textDocument/documentHighlight", Some idNode, Some paramsObj ->
                                LspHandlers.handleDocumentHighlight idNode paramsObj
                            | Some "textDocument/signatureHelp", Some idNode, Some paramsObj ->
                                LspHandlers.handleSignatureHelp idNode paramsObj
                            | Some "textDocument/rename", Some idNode, Some paramsObj ->
                                LspHandlers.handleRename idNode paramsObj
                            | Some "textDocument/prepareRename", Some idNode, Some paramsObj ->
                                LspHandlers.handlePrepareRename idNode paramsObj
                            | Some "workspace/symbol", Some idNode, Some paramsObj ->
                                LspHandlers.handleWorkspaceSymbol idNode paramsObj
                            | Some "textDocument/codeAction", Some idNode, Some paramsObj ->
                                LspHandlers.handleCodeAction idNode paramsObj
                            | Some "textDocument/inlayHint", Some idNode, Some paramsObj ->
                                LspHandlers.handleInlayHints idNode paramsObj
                            | Some "fscript/viewAst", Some idNode, Some paramsObj ->
                                LspHandlers.handleViewAst idNode paramsObj
                            | Some "fscript/viewInferredAst", Some idNode, Some paramsObj ->
                                LspHandlers.handleViewInferredAst idNode paramsObj
                            | Some "fscript/stdlibSource", Some idNode, Some paramsObj ->
                                LspHandlers.handleStdlibSource idNode paramsObj
                            | Some _, Some idNode, _ ->
                                LspProtocol.sendError idNode -32601 "Method not found"
                            | _ -> ()
                with ex ->
                    // Never crash the server loop on malformed input.
                    let p = JsonObject()
                    p["type"] <- JsonValue.Create(1)
                    p["message"] <- JsonValue.Create($"FScript LSP internal error: {ex.Message}")
                    LspProtocol.sendNotification "window/logMessage" (Some p)

File: src/FScript.LanguageServer/LspServer.cs
using System.Text.Json.Nodes;
using Microsoft.FSharp.Core;
using FSLspHandlers = FScript.LanguageServer.LspHandlers;
using FSLspProtocol = FScript.LanguageServer.LspProtocol;

namespace FScript.LanguageServer.CSharp;

internal sealed class LspServer
{
    private bool _shutdownRequested;

    public void Run()
    {
        var input = Console.OpenStandardInput();

        while (true)
        {
            var raw = JsonRpcWire.ReadMessage(input);
            if (raw is null)
            {
                break;
            }

            JsonObject? message;
            try
            {
                message = JsonNode.Parse(raw) as JsonObject;
            }
            catch
            {
                continue;
            }

            if (message is null)
            {
                continue;
            }

            var method = message["method"]?.GetValue<string>();
            var idNode = message["id"];
            var paramsObj = message["params"] as JsonObject;

            if (string.IsNullOrWhiteSpace(method))
            {
                continue;
            }

            try
            {
                if (idNode is null)
                {
                    HandleNotification(method!, paramsObj);
                    if (_shutdownRequested && string.Equals(method, "exit", StringComparison.Ordinal))
                    {
                        break;
                    }
                }
                else
                {
                    HandleRequest(idNode, method!, paramsObj);
                }
            }
            catch (Exception ex)
            {
                var payload = new JsonObject
                {
                    ["type"] = 1,
                    ["message"] = $"FScript LSP internal error (C# host): {ex.Message}"
                };
                FSLspProtocol.sendNotification("window/logMessage", FSharpOption<JsonNode?>.Some(payload));
            }
        }
    }

    private static void HandleNotification(string method, JsonObject? paramsObj)
    {
        switch (method)
        {
            case "initialized":
                break;
            case "textDocument/didOpen":
                if (paramsObj is not null)
                {
                    FSLspHandlers.handleDidOpen(paramsObj);
                }
                break;
            case "textDocument/didChange":
                if (paramsObj is not null)
                {
                    FSLspHandlers.handleDidChange(paramsObj);
                }
                break;
            case "textDocument/didClose":
                if (paramsObj is not null)
                {
                    FSLspHandlers.handleDidClose(paramsObj);
                }
                break;
            case "exit":
                break;
        }
    }

    private void HandleRequest(JsonNode idNode, string method, JsonObject? paramsObj)
    {
        switch (method)
        {
            case "initialize":
                FSLspHandlers.handleInitialize(idNode, paramsObj is null ? FSharpOption<JsonObject>.None : FSharpOption<JsonObject>.Some(paramsObj));
                break;
            case "shutdown":
                _shutdownRequested = true;
                FSLspProtocol.sendResponse(idNode, FSharpOption<JsonNode?>.None);
                break;
            case "textDocument/completion":
                if (paramsObj is not null) FSLspHandlers.handleCompletion(idNode, paramsObj); else SendInvalidParams(idNode);
                break;
            case "textDocument/semanticTokens/full":
                if (paramsObj is not null) FSLspHandlers.handleSemanticTokens(idNode, paramsObj); else SendInvalidParams(idNode);
                break;
            case "textDocument/hover":
                if (paramsObj is not null) FSLspHandlers.handleHover(idNode, paramsObj); else SendInvalidParams(idNode);
                break;
            case "textDocument/definition":
                if (paramsObj is not null) FSLspHandlers.handleDefinition(idNode, paramsObj); else SendInvalidParams(idNode);
                break;
            case "textDocument/typeDefinition":
                if (paramsObj is not null) FSLspHandlers.handleTypeDefinition(idNode, paramsObj); else SendInvalidParams(idNode);
                break;
            case "textDocument/documentSymbol":
                if (paramsObj is not null) FSLspHandlers.handleDocumentSymbol(idNode, paramsObj); else SendInvalidParams(idNode);
                break;
            case "textDocument/references":
                if (paramsObj is not null) FSLspHandlers.handleReferences(idNode, paramsObj); else SendInvalidParams(idNode);
                break;
            case "textDocument/documentHighlight":
                if (paramsObj is not null) FSLspHandlers.handleDocumentHighlight(idNode, paramsObj); else SendInvalidParams(idNode);
                break;
            case "textDocument/signatureHelp":
                if (paramsObj is not null) FSLspHandlers.handleSignatureHelp(idNode, paramsObj); else SendInvalidParams(idNode);
                break;
            case "textDocument/rename":
                if (paramsObj is not null) FSLspHandlers.handleRename(idNode, paramsObj); else SendInvalidParams(idNode);
                break;
            case "textDocument/prepareRename":
                if (paramsObj is not null) FSLspHandlers.handlePrepareRename(idNode, paramsObj); else SendInvalidParams(idNode);
                break;
            case "workspace/symbol":
                if (paramsObj is not null) FSLspHandlers.handleWorkspaceSymbol(idNode, paramsObj); else SendInvalidParams(idNode);
                break;
            case "textDocument/codeAction":
                if (paramsObj is not null) FSLspHandlers.handleCodeAction(idNode, paramsObj); else SendInvalidParams(idNode);
                break;
            case "textDocument/inlayHint":
                if (paramsObj is not null) FSLspHandlers.handleInlayHints(idNode, paramsObj); else SendInvalidParams(idNode);
                break;
            case "fscript/viewAst":
                if (paramsObj is not null) FSLspHandlers.handleViewAst(idNode, paramsObj); else SendInvalidParams(idNode);
                break;
            case "fscript/viewInferredAst":
                if (paramsObj is not null) FSLspHandlers.handleViewInferredAst(idNode, paramsObj); else SendInvalidParams(idNode);
                break;
            default:
                FSLspProtocol.sendError(idNode, -32601, "Method not found");
                break;
        }
    }

    private static void SendInvalidParams(JsonNode idNode)
    {
        FSLspProtocol.sendError(idNode, -32602, "Invalid params");
    }
}
